<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=US-ASCII">
<meta name="generator" content="hevea 2.06">
<style type="text/css">
.li-itemize{margin:1ex 0ex;}
.li-enumerate{margin:1ex 0ex;}
.dd-description{margin:0ex 0ex 1ex 4ex;}
.dt-description{margin:0ex;}
.toc{list-style:none;}
.footnotetext{margin:0ex; padding:0ex;}
div.footnotetext P{margin:0px; text-indent:1em;}
.thefootnotes{text-align:left;margin:0ex;}
.dt-thefootnotes{margin:0em;}
.dd-thefootnotes{margin:0em 0em 0em 2em;}
.footnoterule{margin:1em auto 1em 0px;width:50%;}
.caption{padding-left:2ex; padding-right:2ex; margin-left:auto; margin-right:auto}
.title{margin:2ex auto;text-align:center}
.titlemain{margin:1ex 2ex 2ex 1ex;}
.titlerest{margin:0ex 2ex;}
.center{text-align:center;margin-left:auto;margin-right:auto;}
.flushleft{text-align:left;margin-left:0ex;margin-right:auto;}
.flushright{text-align:right;margin-left:auto;margin-right:0ex;}
div table{margin-left:inherit;margin-right:inherit;margin-bottom:2px;margin-top:2px}
td table{margin:auto;}
table{border-collapse:collapse;}
td{padding:0;}
.cellpadding0 tr td{padding:0;}
.cellpadding1 tr td{padding:1px;}
pre{text-align:left;margin-left:0ex;margin-right:auto;}
blockquote{margin-left:4ex;margin-right:4ex;text-align:left;}
td p{margin:0px;}
.boxed{border:1px solid black}
.textboxed{border:1px solid black}
.vbar{border:none;width:2px;background-color:black;}
.hbar{border:none;height:2px;width:100%;background-color:black;}
.hfill{border:none;height:1px;width:200%;background-color:black;}
.vdisplay{border-collapse:separate;border-spacing:2px;width:auto; empty-cells:show; border:2px solid red;}
.vdcell{white-space:nowrap;padding:0px; border:2px solid green;}
.display{border-collapse:separate;border-spacing:2px;width:auto; border:none;}
.dcell{white-space:nowrap;padding:0px; border:none;}
.dcenter{margin:0ex auto;}
.vdcenter{border:solid #FF8000 2px; margin:0ex auto;}
.minipage{text-align:left; margin-left:0em; margin-right:auto;}
.marginpar{border:solid thin black; width:20%; text-align:left;}
.marginparleft{float:left; margin-left:0ex; margin-right:1ex;}
.marginparright{float:right; margin-left:1ex; margin-right:0ex;}
.theorem{text-align:left;margin:1ex auto 1ex 0ex;}
.part{margin:2ex auto;text-align:center}
</style>
<title>HapCompass Manual
</title>
</head>
<body >
<!--HEVEA command line is: C:\bin\havea\hevea.exe hapcompass_manual.tex -->
<!--CUT STYLE article--><!--CUT DEF section 1 --><table class="title"><tr><td style="padding:1ex"><h1 class="titlemain">HapCompass Manual</h1><h3 class="titlerest">Derek Aguiar<br>
Brown University</h3></td></tr>
</table><!--TOC section id="sec1" Contents-->
<h2 id="sec1" class="section">Contents</h2><!--SEC END --><ul class="toc"><li class="li-toc">
<a href="#sec2">1&#XA0;&#XA0;Introduction</a>
<ul class="toc"><li class="li-toc">
<a href="#sec3">1.1&#XA0;&#XA0;Description</a>
</li><li class="li-toc"><a href="#sec4">1.2&#XA0;&#XA0;Download</a>
</li></ul>
</li><li class="li-toc"><a href="#sec5">2&#XA0;&#XA0;HapCompass options</a>
<ul class="toc"><li class="li-toc">
<a href="#sec6">2.1&#XA0;&#XA0;Usage</a>
</li><li class="li-toc"><a href="#sec7">2.2&#XA0;&#XA0;Arguments</a>
<ul class="toc"><li class="li-toc">
<a href="#sec8">2.2.1&#XA0;&#XA0;Required</a>
</li><li class="li-toc"><a href="#sec9">2.2.2&#XA0;&#XA0;Optional</a>
</li><li class="li-toc"><a href="#sec10">2.2.3&#XA0;&#XA0;HapCompass output</a>
</li></ul>
</li></ul>
</li><li class="li-toc"><a href="#sec11">3&#XA0;&#XA0;File formats</a>
<ul class="toc"><li class="li-toc">
<a href="#sec12">3.1&#XA0;&#XA0;Variants</a>
</li><li class="li-toc"><a href="#sec13">3.2&#XA0;&#XA0;Aligned sequence reads</a>
</li></ul>
</li><li class="li-toc"><a href="#sec14">4&#XA0;&#XA0;Caveats and assumptions</a>
</li><li class="li-toc"><a href="#sec15">5&#XA0;&#XA0;More information</a>
<ul class="toc"><li class="li-toc">
<a href="#sec16">5.1&#XA0;&#XA0;What is haplotype assembly?</a>
</li><li class="li-toc"><a href="#sec17">5.2&#XA0;&#XA0;Author</a>
</li><li class="li-toc"><a href="#sec18">5.3&#XA0;&#XA0;Citations</a>
</li></ul>
</li></ul>
<!--TOC section id="sec2" Introduction-->
<h2 id="sec2" class="section">1&#XA0;&#XA0;Introduction</h2><!--SEC END --><p>HapCompass is a software package for haplotype assembly of diploid, polyploid, and tumor genomes[<a href="#aguiar2012">1</a>,<a href="#Aguiar2013">2</a>,<a href="#aguiar2014">3</a>].</p>
<!--TOC subsection id="sec3" Description-->
<h3 id="sec3" class="subsection">1.1&#XA0;&#XA0;Description</h3><!--SEC END --><p>HapCompass is a graph-theoretic algorithm for genome-wide haplotype assembly.
HapCompass-Polyploidy and HapCompass-Tumor are the first models and algorithms to extend haplotype assembly beyond diploid genomes.</p>
<!--TOC subsection id="sec4" Download-->
<h3 id="sec4" class="subsection">1.2&#XA0;&#XA0;Download</h3><!--SEC END --><p>HapCompass can be downloaded from <a href="http://www.brown.edu/Research/Istrail_Lab/hapcompass.php">http://www.brown.edu/Research/Istrail_Lab/hapcompass.php</a>.</p>
<!--TOC section id="sec5" HapCompass options-->
<h2 id="sec5" class="section">2&#XA0;&#XA0;HapCompass options</h2><!--SEC END -->
<!--TOC subsection id="sec6" Usage-->
<h3 id="sec6" class="subsection">2.1&#XA0;&#XA0;Usage</h3><!--SEC END --><p>usage: java [JVM options] -jar hapcompass.jar [OPTIONS]...<br>

Example usage<br>

java -Xmx1g -jar hapcompass.jar &#X2013;bam example/example.bam &#X2013;vcf example/62_ID0.txt -o example/out</p>
<!--TOC subsection id="sec7" Arguments-->
<h3 id="sec7" class="subsection">2.2&#XA0;&#XA0;Arguments</h3><!--SEC END --><p>The parameters to HAPCOMPASS are given in <code>-&lt;parameter&gt;</code> <code>&lt;value&gt;</code> or <code>--&lt;parameter&gt;=&lt;value&gt;</code> pairs if there are arguments or <code>-&lt;parameter&gt;</code> or <code>--&lt;parameter&gt;</code> for boolean switches.</p>
<!--TOC subsubsection id="sec8" Required-->
<h4 id="sec8" class="subsubsection">2.2.1&#XA0;&#XA0;Required</h4><!--SEC END --><pre class="verbatim">-b,--bam &lt;arg&gt;          The BAM or SAM file to process. The
                        preferred input is a BAM file that is coordinate
                        sorted. If you require name sorted input files, the
                        --namesort option may be used in conjunction
                        with aname sorted BAM or SAM file. If the file is
                        name sorted, mated reads must has the same name.
-o,--output &lt;arg&gt;       Prefix to write output files to. This should
                        include a reference to the directory and a prefix
                        for files, for example: /home/me/myOutput_prefix
 -v,--vcf &lt;arg&gt;         The VCF file VCFFILE giving the variants. The VCF
                        file should follow the VCF 4.1 standard, that
                        is,it should have 8 tab delimited columns: #CHROM
                        POS ID REF ALT QUAL FILTER INFO followed by a
                        FORMAT column and 1 column representing the
                        individual's genotypes.            

</pre>
<!--TOC subsubsection id="sec9" Optional-->
<h4 id="sec9" class="subsubsection">2.2.2&#XA0;&#XA0;Optional</h4><!--SEC END --><pre class="verbatim"> 
 -c,--cache              Enable caching of chunks of the input BAM file.
                         This option is off by default but can be used if
                         you are experiencing problems with the default
                         BAM loading and the BAM file is very large with
                         many very long insert size reads.
 -d,--debug              Turn on debugging. This option will produce an
                         output log file.
 -e,--export             Don't run the algorithm, just export data files.
 -f,--fragment &lt;arg&gt;     Set input fragment filename.
 -g,--downsample &lt;arg&gt;   Downsample an input BAM file.
 -i,--prim               Turn on the Prim-like algorithm for resolution of
                         the general chain graph for polyploid genomes.
 -j,--diverse            Force HapCompass to prefer edges in G_C that
                         include more unique haplotypes. This option
                         sometimes works well when the coverage is low and
                         the sequence read length is long.
 -l,--suffix &lt;arg&gt;       Defines a suffix to trim off read names. This was
                         added as a convinience for processing some files
                         produced by sff_extract.
 -n,--namesort           HapCompass will assume the input SAM/BAM files
                         are sorted by name
 -p,--ploidy &lt;arg&gt;       Set the ploidy of the genome to be assembled.
                         Value must be 2 or higher. Any number above 2
                         will trigger usage of the polyploid algorithm.
                         Default is 2
 -r,--reference &lt;arg&gt;    HapCompass will only produce results from this
                         reference name. This must match a VCF and SAM/BAM
                         reference name.
 -s,--simulate &lt;arg&gt;     HapCompass will simulate reads with properties
                         given by the string:&lt;number of reads&gt;:&lt;read
                         length&gt;:&lt;single base error substitution
                         probability&gt;:&lt;smallest base to sample&gt;:&lt;largest
                         base to sample&gt;:&lt;list of insert sizes seperated
                         by commas&gt;:&lt;list of insert sizes seperated by
                         commas&gt;:&lt;reference name to simulate reads
                         for&gt;:&lt;only produce good reads&gt;Example:
                         1000:200:0.01:1:14000:3500,2500:1150,1250:chr1:tr
                         ue will simulate 1000 paired reads with 200 bases
                         for each read of the pairs, 1% error rate, in the
                         region of [1,14000], with insert sizes 3500 and
                         2500 and insert size standard deviations 1150 and
                         1250 respectively for reference chr1 and only
                         produce reads with 2 or more variants
 -x,--evaluate &lt;arg&gt;     Provide a phasing filename which was output from
                         HapCompass for evaluation.
 -y,--ilp                Output ILP file.
 -z,--iterations &lt;arg&gt;   Number of iterations to run. The larger the value
                         the more accuracte the assembly. Default is 10
                         for ploidy = 2, and 1 for ploidy &gt; 2.
</pre>
<!--TOC subsubsection id="sec10" HapCompass output-->
<h4 id="sec10" class="subsubsection">2.2.3&#XA0;&#XA0;HapCompass output</h4><!--SEC END --><p>Several files will be output to the -o directory.</p><ul class="itemize"><li class="li-itemize">
	<code>OUTPUT_PREFIX_&lt;ALGORITHM&gt;_solution.txt</code><br>
	
	Contains a list of blocks and phasings. A new block is shown by
<code><span style="font-size:small">BLOCK    &lt;start_position_of_block&gt;    &lt;end_position&gt;    &lt;start_snp_number&gt;    &lt;end_snp_number&gt;    &lt;score&gt;</span></code><span style="font-size:medium"><br>
then a list of SNPs in the block...<br>
</span><code><span style="font-size:small">&lt;snp_id&gt;    &lt;snp_position&gt;    &lt;snp_number&gt;    &lt;hap_0_allele&gt;    &lt;hap_1_allele&gt;</span></code><span style="font-size:small"><br>
</span><p><span style="font-size:medium">In the case of assembling more than 2 haplotypes, there will be additional columns for the extra haplotypes.
	</span></p></li><li class="li-itemize"><code><span style="font-size:medium">OUTPUT_PREFIX_reduced_representation.vcf</span></code><span style="font-size:medium"> and<br>
</span><code><span style="font-size:medium">OUTPUT_PREFIX_reduced_representation.sam</span></code><span style="font-size:medium"><br>
	
A list of SAM reads and variants covering the blocks of variants in terms of the variant numbers instead of variant positions.
	</span></li><li class="li-itemize"><code><span style="font-size:medium">OUTPUT_PREFIX_reads.sam</span></code><span style="font-size:medium"><br>
	
A different representation of input reads in terms of true genomic positions.</span><p><span style="font-size:medium">	</span></p></li><li class="li-itemize"><code><span style="font-size:medium">OUTPUT_PREFIX_binning.txt</span></code><span style="font-size:medium"><br>
	
	This file will only be produced if </span><code><span style="font-size:medium">-ploidy &gt; 2</span></code><span style="font-size:medium">. Contains the pairwise variant phasings defined in the compass graph </span><span style="font-size:medium"><span style="font-style:italic">G</span></span><sub><span style="font-size:medium"><span style="font-style:italic">C</span></span></sub><span style="font-size:medium">.
	
	</span></li><li class="li-itemize"><code><span style="font-size:medium">OUTPUT_PREFIX_frags.txt</span></code><span style="font-size:medium"><br>
	
	BAM input files are converted into an alternative representation that contains only the information required by HapCompass.
	This file will only contain the sequence reads that contain 2 or more variant alleles.
</span></li></ul>
<!--TOC section id="sec11" File formats-->
<h2 id="sec11" class="section">3&#XA0;&#XA0;File formats</h2><!--SEC END --><p>The HapCompass package includes sample files for all types of input.</p>
<!--TOC subsection id="sec12" Variants-->
<h3 id="sec12" class="subsection">3.1&#XA0;&#XA0;Variants</h3><!--SEC END --><p>
HapCompass always requires specification of variants in the form of a VCF file.
VCF specification can be found <a href="https://github.com/amarcket/vcf_spec">https://github.com/amarcket/vcf_spec</a>.</p>
<!--TOC subsection id="sec13" Aligned sequence reads-->
<h3 id="sec13" class="subsection">3.2&#XA0;&#XA0;Aligned sequence reads</h3><!--SEC END --><p>There are two options for inputting sequence reads: BAM and fragment files.</p><p>A SAM file is a tab delimited file for specifying sequence reads and their properties. 
A BAM file is the SAM files binary version.
BAM and SAM files may be manipulated using SAMTools.
The BAM specification can be found <a href="http://samtools.sourceforge.net/SAMv1.pdf">http://samtools.sourceforge.net/SAMv1.pdf</a>.</p><p>Fragment files are a <span style="font-weight:bold">tab</span> delimited file format with reads separated by new lines.
The columns are specified as follows: 
</p><dl class="description"><dt class="dt-description">
	<span style="font-weight:bold">column 1:</span></dt><dd class="dd-description"> number of read parts, e.g. 1 for a single read, 2 for two reads (paired sequencing when both ends are informative)
	</dd><dt class="dt-description"><span style="font-weight:bold">column 2:</span></dt><dd class="dd-description"> read name
	</dd><dt class="dt-description"><span style="font-weight:bold">column 3 and on:</span></dt><dd class="dd-description"> the read parts. The number of read parts is defined in column 1. 
</dd></dl><p>
A read part is a <span style="font-weight:bold">space</span> delimited list beginning with a reference name and then a set of one or more position(s), allele(s), base quality score(s) trios. 
The position is the 0-based position of the variant in the VCF file. E.g. the first variant is position 0, the second variant is position 1, and so on.
The allele is the numerically encoded allele, 0 for reference, 1 for first alternative allele, and so on.
The base quality score is the ASCII of Phred-scaled base quality+33.
An example of a paired read is given below <code>&lt;tab&gt;</code> denotes a tab character:
</p><pre class="verbatim">2&lt;tab&gt;sequence_read_1&lt;tab&gt;chr1 0 1 D 1 1 D&lt;tab&gt;chr1 4 0 D 5 0 D 6 0 D
</pre>
<!--TOC section id="sec14" Caveats and assumptions-->
<h2 id="sec14" class="section">4&#XA0;&#XA0;Caveats and assumptions</h2><!--SEC END --><p>HapCompass, and haplotype assembly algorithms in general, requires a set of aligned sequence reads and variants calls.
The polyploid algorithm requires knowledge of the number of homologous copies.</p>
<!--TOC section id="sec15" More information-->
<h2 id="sec15" class="section">5&#XA0;&#XA0;More information</h2><!--SEC END -->
<!--TOC subsection id="sec16" What is haplotype assembly?-->
<h3 id="sec16" class="subsection">5.1&#XA0;&#XA0;What is haplotype assembly?</h3><!--SEC END --><p>Standard genome sequencing workflows produce contiguous DNA segments of an unknown chromosomal origin.
<span style="font-style:italic">De-novo</span> assemblies for genomes with two sets of chromosomes (<span style="font-style:italic">diploid</span>) or more (<span style="font-style:italic">polyploid</span>) produce consensus sequences in which the relative haplotype phase between variants is undetermined.
The set of sequencing reads can be mapped to the phase-ambiguous reference genome and the diploid chromosome origin can be determined but, without knowledge of the haplotype sequences, reads cannot be mapped to the particular haploid chromosome sequence.
However, sequence reads are derived from a single haploid fragment and thus provide valuable phase information when they contain two or more variants.
The <span style="font-style:italic">haplotype assembly problem</span> aims to compute the haplotype sequences for each chromosome given a set of aligned sequence reads to the genome and variant information.
The haplotype phase of variants is inferred from assembling overlapping sequence reads.</p>
<!--TOC subsection id="sec17" Author-->
<h3 id="sec17" class="subsection">5.2&#XA0;&#XA0;Author</h3><!--SEC END --><p><a href="www.derekaguiar.com">Derek Aguiar</a> earned his PhD from the <a href="http://cs.brown.edu/">Department of Computer Science</a> at <a href="http://brown.edu/">Brown University</a> working under the direction of <a href="http://www.brown.edu/Research/Istrail_Lab/sorin.php">Sorin Istrail</a>.</p>
<!--TOC subsection id="sec18" Citations-->
<h3 id="sec18" class="subsection">5.3&#XA0;&#XA0;Citations</h3><!--SEC END --><ol class="enumerate" type=1><li class="li-enumerate">
	HapCompass for diploid genomes[<a href="#aguiar2012">1</a>].
	</li><li class="li-enumerate">HapCompass for tumor and polyploid genomes[<a href="#Aguiar2013">2</a>,<a href="#aguiar2014">3</a>].
</li></ol><!--TOC section id="sec19" References-->
<h2 id="sec19" class="section">References</h2><!--SEC END --><dl class="thebibliography"><dt class="dt-thebibliography">
<a id="aguiar2012"><span style="color:purple">[1]</span></a></dt><dd class="dd-thebibliography">
Derek Aguiar and Sorin Istrail,
<em>HAPCOMPASS: A fast cycle basis algorithm for accurate haplotype assembly
	of sequence data</em>.
	J. Comput. Biol.,
<span style="font-weight:bold">19</span>, 2013</dd><dt class="dt-thebibliography"><a id="Aguiar2013"><span style="color:purple">[2]</span></a></dt><dd class="dd-thebibliography">
Derek Aguiar and Sorin Istrail,
<em>Haplotype assembly in polyploid genomes and identical by descent shared tracts</em>.
	Bioinformatics,
Pacific Symposium of Biocomputing,
<span style="font-weight:bold">29</span>:352-360, 2013</dd><dt class="dt-thebibliography"><a id="aguiar2014"><span style="color:purple">[3]</span></a></dt><dd class="dd-thebibliography">
Derek Aguiar, Wendy S.W. Wong and Sorin Istrail,
<em>Tumor haplotype assembly algorithms for cancer genomics</em>.
Pacific Symposium of Biocomputing,
2014</dd></dl><!--CUT END -->
<!--HTMLFOOT-->
<!--ENDHTML-->
<!--FOOTER-->
<hr style="height:2"><blockquote class="quote"><em>This document was translated from L<sup>A</sup>T<sub>E</sub>X by
</em><a href="http://hevea.inria.fr/index.html"><em>H</em><em><span style="font-size:small"><sup>E</sup></span></em><em>V</em><em><span style="font-size:small"><sup>E</sup></span></em><em>A</em></a><em>.</em></blockquote></body>
</html>
